Move lua event pixelpipe-processing-complete to color_picker_proxy

wpferguson opened the pull request:
It seems that _color_picker_proxy_preview_pipe_callback() is called after all pixelpipe processing is complete, so I moved the lua event trigger for `pixelpipe-processing-complete` there.  This allows scripts using darktable.gui.action() to run without having to resort to adding sleeps to allow the pixelpipe to complete before running the next action.

I've attached a script that adds a button to selected images named `display in darkroom`.   Select a set of images in lighttable view, then click the button.  Each image should be loaded and 0.5 ev of exposure added to it.
[display_in_darkroom.zip](https://github.com/darktable-org/darktable/files/10833032/display_in_darkroom.zip)

Fixes #13745

dterrahe commented:
Have you tested if it makes a difference whether you trigger the lua "pixelpipe-processing-complete" in `_color_picker_proxy_preview_pipe_callback`, after _catching_ the `DT_SIGNAL_DEVELOP_PREVIEW_PIPE_FINISHED` signal or in `dt_dev_process_preview_job` just after _throwing_ the signal (https://github.com/darktable-org/darktable/blob/ee15dec0e5cfc54b96f1e9365ce5d3b812e7356d/src/develop/develop.c#L372)? It would make somewhat more logical sense to me to throw both from the same location. If there was a difference, I would like to understand why, to understand if there is any random chance of failure involved.

wpferguson commented:
I did try catching it there.

I changed pixelpipe-processing-complete to image-pixelpipe-processing-complete and then added preview-pixelpipe-processing-complete just after line 372 in develop.c.

When I tested sometimes image fired first, then preview and other times it was reversed.  They always fired together except for spot exposure measurement which only runs in the preview pipe.  When I was testing the tone equalizer masking (a 3 step process) the first step would start, the events would fire, then the second step, then the events, then the third step, then the events.  The GUI was still on the first step, so it hadn't completed whatever it was doing.  Then I watched as the other two steps completed (they acted as if they were queued up for execution).

Moving the event to _color_picker_proxy_preview_pipe_callback lets each step complete before the next is started.  I poll every 350ms to see if the event has fired.  With the event in _color_picker_proxy_preview_pipe_callback  it took 2 checks during the mask calculations to complete each step, which I believe is much more accurate then when I was firing the events from develop.c.

I don't really understand how _color_picker_proxy_preview_pipe_callback manages to wait until everything is complete but it seems to and this seems to be the best place to fire the event from.  I also don't understand how you found it, but I'm glad you did.

I guess the acid test is to try the script above on 4.2.1 and then try it with this patch.

wpferguson commented:
@TurboGit 8 hours ago I decided to put the WIP label on this but I didn't have a chance until now and I found you've done.  It seems line one of us is reading the other's mind :smiley:

wpferguson commented:
Did some more research and this is what I came up with:

- There was an error in my event handling in the script, which caused most of the problems.  I created a situation where the complete_flag was always true, so the script never waited on the pixelpipe to finish. 
- Once I fixed that, I decided to test the various placements of the pixelpipe event.  I chose:
1. src/develop/develop.c:660 - where it currently is in master.  I called it image-pixelpipe-processing-complete for the test
2. src/develop/develop.c:373 - at the end of the preview pipe.  I called it preview-pixelpipe-processing-complete for the test.
3. src/gui/color_picker_proxy.c:323 - the proposed new location.  I called it picker-pixelpipe-processing-complete for the test.
- I ran a test using the preview-pixelpipe-processing-complete event to determine when it was safe to proceed.  The other two events were caught and a message printed so that we could see where the events fired in relation to each other.  The test consisted of selecting 20 random play raw images from pixls.us and processing them in a loop.  Each image is opened in darkroom view, the current exposure setting determined, 0.5 EV added to it and then applied to the image.
- Here's some sample output from the test
```
    55.6420 LUA changing image to 20220807_0009.NEF
    55.8678 LUA loaded
    55.8678 LUA resetting loaded flag
    55.8678 LUA waiting for complete_flag
    55.8682 LUA caught darkroom-image-loaded event with clean = true
    55.9679 LUA waiting for complete_flag
    56.0202 LUA caught image-pixelpipe-processing-complete
    56.0681 LUA waiting for complete_flag
    56.1588 LUA caught preview-pixelpipe-processing-complete
    56.1613 LUA caught picker-pixelpipe-processing-complete
    56.1683 LUA complete
    56.1683 LUA resetting complete_flag
    56.1683 LUA calling process_image
    56.1683 LUA in rel_adjust
    56.1815 LUA status is 0.70000004768372 and adjusted value should be 1.2000000476837
    56.1926 LUA waiting for complete_flag
    56.2707 LUA caught image-pixelpipe-processing-complete
    56.2928 LUA waiting for complete_flag
    56.3055 LUA caught preview-pixelpipe-processing-complete
    56.3057 LUA caught picker-pixelpipe-processing-complete
    56.3930 LUA complete
    56.3930 LUA resetting complete_flag
    56.3931 LUA loading next image


    56.3931 LUA changing image to 20220808_0103.CR2
    56.5909 LUA loaded
    56.5909 LUA resetting loaded flag
    56.5909 LUA waiting for complete_flag
    56.5913 LUA caught darkroom-image-loaded event with clean = true
    56.6910 LUA waiting for complete_flag
    56.7177 LUA caught image-pixelpipe-processing-complete
    56.7912 LUA waiting for complete_flag
    56.8122 LUA caught preview-pixelpipe-processing-complete
    56.8147 LUA caught picker-pixelpipe-processing-complete
    56.8913 LUA complete
    56.8914 LUA resetting complete_flag
    56.8914 LUA calling process_image
    56.8915 LUA in rel_adjust
    56.8917 LUA status is 0.70000004768372 and adjusted value should be 1.2000000476837
    56.9139 LUA waiting for complete_flag
    56.9817 LUA caught preview-pixelpipe-processing-complete
    56.9818 LUA caught picker-pixelpipe-processing-complete
    57.0141 LUA complete
    57.0141 LUA resetting complete_flag
    57.0141 LUA loading next image

```

I ran several tests on different batches of images, with the total processing time from lighttable view to libhttable view being between 12 and 15 seconds. 

###Conclusions

1. The` image-pixelpipe-...` event always fires first, followed by `preview-pixelpipe-...`, then `picker-pixelpipe-...`.
2. The `image-pixelpipe-processing-complete` event does not cat all pixelpipe processing.  Spot exposure measurement only runs in the preview pixelpipipe, so that means in order to catch all processing we need to use either `preview-pixelpipe-processing-complete` or `picker-pixelpipe-processing-complete`.
3. `picker-pixelpipe-processing-complete` always fires last, so it gives the most time for the UI to be ready for the next input.  That makes it the "safest".
4. No errors were encountered using `preview-pixelpipe-processing-complete`.  I tested a lot of scenarios, especially the ones that gave me fits such as tone equalizer mask building, using both `image-pixelpipe-processing-complete` and `preview-pixelpipe-processing-complete` and didn't encounter any errors.

### Plan

- Close this PR
- Create a new PR moving the `pixelpipe-processing-complete` event in src/develop/develop.c from the end of the image pipe to the end of the preview pipe (src/develop/develop.c:372).

@dterrahe, @TurboGit thoughts?

wpferguson commented:
Superseded by #13854
